lib/filterx/func-str-transform.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2024 Axoflow
+ * Copyright (c) 2024 Attila Szakacs
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filterx/func-str-transform.h"
+#include "filterx/object-string.h"
+#include "filterx/filterx-eval.h"
+
+static const gchar *
+_extract_str_arg(FilterXExpr *s, GPtrArray *args, gssize *len)
+{
+  if (args == NULL || args->len != 1)
+    {
+      filterx_simple_function_argument_error(s, "Requires exactly one argument", FALSE);
+      return NULL;
+    }
+
+  gsize inner_len;
+  FilterXObject *object = g_ptr_array_index(args, 0);
+  const gchar *str = filterx_string_get_value(object, &inner_len);
+  if (!str)
+    {
+      filterx_simple_function_argument_error(s, "Object must be string", FALSE);
+      return NULL;
+    }
+
+  *len = (gssize) MIN(inner_len, G_MAXINT64);
+  return str;
+}
+
+FilterXObject *
+filterx_simple_function_lower(FilterXExpr *s, GPtrArray *args)
+{
+  gssize len;
+  const gchar *str = _extract_str_arg(s, args, &len);
+  if (!str)
+    return NULL;
+
+  gchar *lower = g_ascii_strdown(str, len);
```

Comment:
I'd try to avoid allocations here if possible.

Reply:
From `g_utf8_strup()`:
> The exact manner that this is done depends on the current locale, and may result in the number of characters in the string increasing. (For instance, the German ess-zet will be changed to SS.)

The only way I see to optimize this would be to use `ScratchBuffers` and a `GString` API that converts the string in-place (we would save the allocation of the new string with `ScratchBuffers`, but would still need to copy), but there is none for UTF-8 unfortunately.

I am open for ideas, if there's any.

Reply:
Check if the size of the string is less than 256 chars and use a local buffer if it is.

But let's see if this shows up in our performance test.

Reply:
The problem is that for UTF-8 there is no in-place replace function. For ASCII we have one in `GString::`.